src/metabase/util/analytics.clj

```diff
@@ -0,0 +1,115 @@
+(ns metabase.util.analytics
+  "Functions for sending Snowplow analytics events"
+  (:require [clojure.core.async :as a]
+            [clojure.tools.logging :as log]
+            [metabase.config :as config]
+            [metabase.models.setting :refer [defsetting]]
+            [metabase.public-settings :as public-settings]
+            [metabase.util.i18n :as i18n :refer [deferred-tru trs]])
+  (:import [com.snowplowanalytics.snowplow.tracker Subject$SubjectBuilder Tracker Tracker$TrackerBuilder]
+           [com.snowplowanalytics.snowplow.tracker.emitter BatchEmitter BatchEmitter$Builder Emitter]
+           [com.snowplowanalytics.snowplow.tracker.events Unstructured Unstructured$Builder]
+           [com.snowplowanalytics.snowplow.tracker.http ApacheHttpClientAdapter ApacheHttpClientAdapter$Builder]
+           com.snowplowanalytics.snowplow.tracker.payload.SelfDescribingJson
+           org.apache.http.client.HttpClient
+           [org.apache.http.impl.client CloseableHttpClient HttpClientBuilder HttpClients]
+           org.apache.http.impl.conn.PoolingHttpClientConnectionManager))
+
+(defsetting snowplow-url
+  (deferred-tru "The URL of the Snowplow collector to send analytics events to")
+  :default (if config/is-prod?
+             "https://sp.metabase.com"
+             ;; See the iglu-schema-registry repo for instructions on how to run Snowplow Micro locally for development
+             "http://localhost:9095")
+  :visibility :public)
+
+(def ^:private ^Emitter emitter
+  "An instance of a Snowplow emitter"
+  (delay
+   (let [^PoolingHttpClientConnectionManager manager (new PoolingHttpClientConnectionManager)
+         ^HttpClientBuilder client-builder (. ^HttpClients HttpClients custom)
+         ^HttpClient client (-> client-builder
+                                (.setConnectionManager manager)
+                                (.build))
+         ^ApacheHttpClientAdapter$Builder adapter-builder (-> (. ^ApacheHttpClientAdapter ApacheHttpClientAdapter builder)
+                                                              (.httpClient ^CloseableHttpClient client)
+                                                              (.url (snowplow-url)))
+         ^ApacheHttpClientAdapter adapter (.build adapter-builder)
+         ^BatchEmitter$Builder batch-emitter-builder (-> (. ^BatchEmitter BatchEmitter builder)
+                                                         (.bufferSize 1)
+                                                         (.httpClientAdapter adapter))]
```

Comment:
This is super ugly and hard to understand; is there a cleaner way to write it or is this just inherent to java interop code?

Reply:
```clojure
(def ^:private ^Emitter emitter
  "An instance of a Snowplow emitter"
  (delay
    (let [client (-> (HttpClients/custom)
                     (.setConnectionManager (PoolingHttpClientConnectionManager.))
                     (.build))
          builder (-> (ApacheHttpClientAdapter/builder)
                      (.httpClient client)
                      (.url (snowplow-url)))
          adapter (.build ^ApacheHttpClientAdapter$Builder builder)
          batch-emitter-builder (-> (BatchEmitter/builder)
                                    (.bufferSize 1)
                                    (.httpClientAdapter adapter))]
      (.build ^BatchEmitter$Builder batch-emitter-builder))))
```
Couple things:
- static methods can be done as `(BatchEmitter/builder)` rather than `(. ^BatchEmitter BatchEmitter builder)`. Its a call on the class so you no longer have to typehint it.
- There's some nasty bit of inheritance here which causes the builders to need to be bound. For instance, the `BatchEmitter$Builder` has a method `bufferSize` but inherits the `httpClientAdapter` from somewhere else. So we can leave those as is, but we can choose where to hint. It gets a bit more readable when the hint is on the one call-site

Reply:
I think this makes it more readable as well: we need a polled http client and an adapter to hook that up to the emitter.

Reply:
@dpsutton thanks, this is great. I written much java interop code yet so I was just trying to figure it out as I went